node/bft/src/primary.rs

```diff
@@ -1485,7 +1491,17 @@ impl<N: Network> Primary<N> {
             // If we do not have the certificate, request it.
             if !self.storage.contains_certificate(*certificate_id) {
                 trace!("Primary - Found a new certificate ID for round {round} from '{peer_ip}'");
-                // TODO (howardwu): Limit the number of open requests we send to a peer.
+                let mut outstanding_requests = self.outstanding_certificate_requests.lock();
+                let entry = outstanding_requests.entry((peer_ip, round));
+                // Get the number of outstanding requests for the peer. Note: the map is cleared on each round.
+                let num_outstanding_requests = entry.and_modify(|value| *value += 1).or_insert(1);
+                if *num_outstanding_requests > MAX_OUTSTANDING_CERTIFICATE_REQUESTS {
+                    return Err(anyhow!(
```

Comment:
If we return here and there are already futures in `fetch_certificates`, those will cancelled. In other words, if `fetch_missing_certificates` gets a list of more than 20 certificates, none of them will be fetched and we will not be able to fetch any certificates from this peer for this round.

Is that the desired behavior? If so, we should expand the function documentation to state that, especially if there are cases where a benign/honest validator sends more than 20 requests.